mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs

```diff
@@ -0,0 +1,219 @@
+use crate::z_coin::storage::{scan_cached_block, validate_chain, BlockDbImpl, BlockProcessingMode, CompactBlockRow};
+use crate::z_coin::z_coin_errors::ZcoinStorageError;
+use crate::z_coin::ZcoinConsensusParams;
+
+use common::async_blocking;
+use db_common::sqlite::rusqlite::{params, Connection};
+use db_common::sqlite::{query_single_row, run_optimization_pragmas, rusqlite};
+use itertools::Itertools;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::*;
+use protobuf::Message;
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+use zcash_client_backend::data_api::error::Error as ChainError;
+use zcash_client_backend::proto::compact_formats::CompactBlock;
+use zcash_client_sqlite::error::{SqliteClientError as ZcashClientError, SqliteClientError};
+use zcash_extras::NoteId;
+use zcash_extras::WalletRead;
+use zcash_primitives::block::BlockHash;
+use zcash_primitives::consensus::BlockHeight;
+
+impl From<ZcashClientError> for ZcoinStorageError {
+    fn from(value: ZcashClientError) -> Self {
+        match value {
+            SqliteClientError::CorruptedData(err) => Self::CorruptedData(err),
+            SqliteClientError::IncorrectHrpExtFvk => Self::IncorrectHrpExtFvk,
+            SqliteClientError::InvalidNote => Self::InvalidNote(value.to_string()),
+            SqliteClientError::InvalidNoteId => Self::InvalidNoteId,
+            SqliteClientError::TableNotEmpty => Self::TableNotEmpty(value.to_string()),
+            SqliteClientError::Bech32(err) => Self::DecodingError(err.to_string()),
+            SqliteClientError::Base58(err) => Self::DecodingError(err.to_string()),
+            SqliteClientError::DbError(err) => Self::DecodingError(err.to_string()),
+            SqliteClientError::Io(err) => Self::IoError(err.to_string()),
+            SqliteClientError::InvalidMemo(err) => Self::InvalidMemo(err.to_string()),
+            SqliteClientError::BackendError(err) => Self::BackendError(err.to_string()),
+        }
+    }
+}
+
+impl From<ChainError<NoteId>> for ZcoinStorageError {
+    fn from(value: ChainError<NoteId>) -> Self { Self::SqliteError(ZcashClientError::from(value)) }
+}
+
+impl BlockDbImpl {
+    #[cfg(all(not(test)))]
+    pub async fn new(_ctx: MmArc, ticker: String, path: Option<impl AsRef<Path>>) -> MmResult<Self, ZcoinStorageError> {
```

Comment:
why does path have to be an option?

Reply:
it's not needed for WASM only sqlite

Reply:
But I see that `None` is never used, the path always `Some` in all uses of this function.

Reply:
refactored.